#105 Add non-const functions to map and unordered_map that allow value modification

peter-moran opened the pull request:
As discussed in #105, this library could support `map`s and `unordered_map`s where the keys are immutable and chosen at compile time, but the values are mutable.

This PR accomplishes this by adding non-const implementations of functions in `map` and `unordered_map`.

To avoid code duplication, I made template functions that the const and non-const functions can call to. By doing this, the compiler can generate the const and non-const implementation from one template.

Additional tests, documentation, and examples should be added with this PR, but I was hoping to get some implementation feedback for now.

serge-sans-paille reviewed (commented):
Thanks for the great PR! I don't see any significant issue with it, minor nits / discussion points inline.

peter-moran commented on examples/value_modification.cpp:
```diff
@@ -0,0 +1,29 @@
+
+int main() {
+  // To make a frozen::unordered_map where you can modify the values, make a
+  // non-constexpr instance:
```
[Just looked that up](https://en.cppreference.com/w/cpp/language/constinit). Very neat, and definitely something to recommend in this case so you can guarantee the work of hashing the keys is done in compile time!

peter-moran replied:
Added

peter-moran commented on examples/value_modification.cpp:
```diff
@@ -0,0 +1,29 @@
+
+  // Find also works
+  auto found = fruits.find("n_oranges");
+  if (found == fruits.end()) {
```
Added to example

peter-moran commented on include/frozen/unordered_map.h:
```diff
@@ -146,8 +146,60 @@ class unordered_map {
+  }
+
+  /*
+   * Template used to reduce code duplication between const and non-const
```
I'll remove it

peter-moran replied:
Makes sense

peter-moran commented on include/frozen/map.h:
```diff
@@ -141,40 +145,91 @@ class map {
-  constexpr const_iterator upper_bound(Key const &key) const {
-    auto const where = bits::lower_bound<N>(items_.begin(), key, less_than_);
-    if ((where != end()) && !less_than_(key, *where))
+  /*
```
I'll remove it

peter-moran commented on include/frozen/unordered_map.h:
```diff
@@ -146,8 +146,60 @@ class unordered_map {
+    auto& kv = self.lookup(key);
+    using kv_ptr = decltype(&kv);
+    if (self.equal_(kv.first, key))
+      return std::pair<kv_ptr, kv_ptr>{&kv, &kv + 1};
```
I agree, I far prefer that syntax, but I cannot get it to work.

The issue is that the return value changes if `This` is const pointer or not. Doing `return {&kv, &kv + 1}` results in the compiler complaining it cannot determine the return type.

Trailing return types also do not work, since the return type is based on `decltype(&kv)` and `kv` cannot be used in that context.

peter-moran commented:
Should "just" need tests now

peter-moran commented on examples/value_modification.cpp:
```diff
@@ -0,0 +1,29 @@
+
+  // Range also works
+  auto range = fruits.equal_range("n_apples");
+}
```
Tests added

peter-moran commented:
@serge-sans-paille should be good to re-review. I've addressed all your comments and added new tests which should have full coverage of the new non-const methods of map and unordered map.

If that all looks good I'd like to wrap up by adding a section to the readme explaining this feature.

peter-moran commented:
@serge-sans-paille should be ready for a final review. I've updated the docs to explain the new functionality.

serge-sans-paille commented:
Thanks!
